source/io.c

```diff
@@ -698,25 +698,25 @@ PyObject *aws_py_tls_connection_options_set_server_name(PyObject *self, PyObject
 /* aws_input_stream implementation for accessing Python I/O classes */
 struct aws_input_stream_py_impl {
     struct aws_input_stream base;
+    struct aws_allocator *allocator;
 
     bool is_end_of_stream;
 
     /* Weak reference proxy to python self. */
     PyObject *self_proxy;
```

Comment:
"Weak reference proxy to python self"
😬

We got some real bad lifetime management going on here, due to the fact that none of this stuff used to be refcounted, and now's the time to unwind it.

Lifetime management currently works like this:
`python class awscrt.io.InputStream`
* keeps alive `python class IOStream`
* keeps alive `C struct aws_input_stream_py_impl`

`C struct aws_input_stream_py_impl`
* only has a **WEAK** reference to its own python class, assuming the python class can't go away before its `_binding`
* and accomplish this by painfully keeping the python class alive everywhere that uses it

but now that the C input stream can be kept alive longer than the python class, we're in trouble

We need to keep python stuff alive as long as the C part is alive, but can't make a circular reference.

I have ideas.... let's chat. too slow to write down